homework_9/hw9.py

```diff
@@ -0,0 +1,248 @@
+Message('если бы ты села делать раньше, тебе было бы легче', Valeri).send(telegram)
+Message('мы не можем быть больше друзьями,  прощай!', Kate).send(telegram)
+telegram.show_chat()
+
+telegram.show_last_message()
+
+#Задание 2
+
+class Args:
+    
+    def __init__(self, *args, **kwargs):
+        self._args = args
+        self._kwargs = kwargs 
+        
+    def __rlshift__(self, other):
+        return other(*self._args, **self._kwargs)
+
+sum << Args([1, 2])
+(lambda a, b, c: a**2 + b + c) << Args(1, 2, c=50)
+
+# Задание 3
+
+class StrangeFloat(float):
+    
+    def __getattr__(self, fun):
+        
+        action, number = fun.split('_')
+        
+        if action == 'add':
+            return StrangeFloat(self + float(number))
+        elif action == 'subtract':
+            return StrangeFloat(self - float(number))
+        elif action == 'multiply':
+            return StrangeFloat(self * float(number))
+        elif action == 'divide':
+            return StrangeFloat(self / float(number))
+        else:
+            raise AttributeError
+
+# Задание 4
+import numpy as np
+
+matrix = []
+for idx in range(0, 100, 10):
+    matrix.__iadd__([list(range(idx, idx.__add__(10)))])
+    
+selected_columns_indices = list(filter(lambda x: x in range(1, 5, 2), range(matrix.__len__())))
+selected_columns = map(lambda x: [x[col] for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr.__getitem__((slice.__call__(None), 1)).__mod__(3).__eq__(0)
+new_arr = arr.__getitem__(mask)
+
+product = new_arr.__matmul__(new_arr.T)
+
+if (product[0].__lt__(1000)).all() and (product[2].__gt__(1000)).any():
+    print(product.mean())
+
+# Задание 5
+
+from abc import ABC, abstractmethod
+
+class BiologicalSequence(ABC):
+    
+    @abstractmethod
+    def __len__(self):
+        pass
+
+    @abstractmethod
+    def __getitem__(self, index):
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        pass
+
+    @abstractmethod
+    def check_seq(self):
+         pass
+        
+        
+class NucleicAcidSequence(BiologicalSequence):
+    
+    nucleotides_compl = {'a' : 't',
+                         't' : 'a',
+                         'g' : 'c',
+                         'c' : 'g',
+                         'u' : 'a'}
+    nucls = 'agcut'
+    
+    def __init__(self, seq):
+        self.seq = seq.lower()
+    
+    def __len__(self):
+        return len(self.seq)
+
+    def __getitem__(self, index):
+        return self.seq[index]
+
+    def __str__(self):
+        return str(self.seq)
+
+    def check_seq(self):
+         
+        if set(self.seq) <= set(self.nucls):
+            return True
+        
+        else:
+            return False
+        
+    def complement(self):
+        
+        if type(self) == NucleicAcidSequence:
+            raise NotImplementedError
+            
+        elif self.check_seq():
+            return ''.join([self.nucleotides_compl[nucl] for nucl in self.seq])
```

Comment:
При комплементировании мы ожидаем получить данные такого же типа, а не строку. Чинится это вот так (по сути это как в 3 задании)

```suggestion
            return type(self)(''.join([self.nucleotides_compl[nucl] for nucl in self.seq]))
```

При комплементировании ДНК мы получим ДНК и так для любого типа